primitives/src/merkle_tree/mod.rs

```diff
@@ -241,11 +240,25 @@ pub trait MerkleTreeScheme: Sized {
     //     pos: impl Iterator<Item = usize>,
     //     proof: impl Borrow<Self::BatchProof>,
     // ) -> Result<(), PrimitivesError>;
+
+    /// Return an iterator that iterates through all element that are not
+    /// forgetton
+    fn iter(&self) -> MerkleTreeIter<Self::Element, Self::Index, Self::NodeValue>;
 }
 
 /// Merkle tree that allows insertion at back. Abstracted as a commitment for
 /// append-only vector.
 pub trait AppendableMerkleTreeScheme: MerkleTreeScheme {
+    /// Construct a new Merkle tree with given height from a data slice
+    /// * `height` - height of the Merkle tree, if `None`, it will calculate the
+    ///   minimum height that could hold all elements.
+    /// * `elems` - an iterator to all elements
+    /// * `returns` - A constructed Merkle tree, or `Err()` if errors
+    fn from_elems(
+        height: Option<usize>,
+        elems: impl IntoIterator<Item = impl Borrow<Self::Element>>,
```

Comment:
since we are going to compute the required `height` regardless of if it's None or Some, why ask? (are we meant to indicate an upperbound `capacity`?)

Reply:
Yes for capacity.
You could start with some less elements and eventually have a large tree